lib/ddq-backend-mysql.js

```diff
@@ -0,0 +1,386 @@
+        );
+    }
+
+
+    /**
+     * Deletes a specific record from the database. If unsuccessful, this
+     * will call the command to requeue the record.
+     *
+     * @param {Object} ddqBackendInstance
+     * @param {string} recordId
+     */
+    function remove(ddqBackendInstance, recordId) {
+        ddqBackendInstance.connection.query(`DELETE FROM ??
+            WHERE hash = ?
+            AND requeued = false;`,
+            [
+                ddqBackendInstance.config.table,
+                recordId
+            ],
+            (err) => {
+                if (err) {
+                    console.error("Error deleting record with id:", recordId);
+                    console.error("Attempting requeue");
+                    requeue(ddqBackendInstance, recordId, (requeueErr) => {
+                        ddqBackendInstance.emit(EMIT_ERR, requeueErr);
+                    });
+                }
+            }
+        );
+    }
+
+
+    /**
+     * Ensures that heartbeatCleanup is ready to run according to the
+     * heartbeatCleanupDelay.
+     *
+     * @param {Object} ddqBackendInstance
+     */
+    function restore(ddqBackendInstance) {
+        /**
+         * Queries for jobs that have stopped, and primes them to be cleaned up.
+         */
+        function heartbeatCleanup() {
+            ddqBackendInstance.connection.query(`UPDATE ??
+                SET isProcessing = false, owner = null, requeued = false
+                WHERE DATEDIFF(NOW(), SEC_TO_TIME(?)) > heartbeatDate
+                AND isProcessing = true;`,
+                [
+                    ddqBackendInstance.config.table,
+                    ddqBackendInstance.config.heartbeatLifetimeSeconds
+                ],
+                (err) => {
+                    if (err) {
+                        ddqBackendInstance.emit(EMIT_ERR, err);
+                    }
+                }
+            );
+
+            ddqBackendInstance.restorer = timers.setTimeout(heartbeatCleanup,
+            ddqBackendInstance.heartbeatCleanupDelayMs);
+        }
+
+        ddqBackendInstance.restorer = timers.setTimeout(heartbeatCleanup,
+            ddqBackendInstance.heartbeatCleanupDelayMs);
+    }
+
+
+    /**
+     * Class for the ddqBackendMySql. Inherits from EventEmitter.
+     */
+    class DdqBackendMySql extends EventEmitter {
+        /**
+         * Default DdqBackendMySql properties.
+         *
+         * @param {Object} backendConfig
+         */
+        constructor() {
+            configValidation.validateConfig(backendConfig);
+            super();
+            this.config = backendConfig;
+            this.connection = null;
+            this.currentlyPolling = false;
+            this.owner = crypto.randomBytes(64).toString("hex");
+            this.poller = null;
+            this.restorer = null;
+            this.topics = backendConfig.topics;
+        }
+
+
+        /**
+         * Queries the database to get a random record. Returns that random
```

Comment:
This emits and does not return.

This should probably take a callback to indicate when the checking is complete so you can schedule more things to occur *after* the check finishes as opposed to in tandem with the check.